firebase-inappmessaging/src/main/java/com/google/firebase/inappmessaging/internal/MetricsLoggerClient.java

```diff
@@ -83,78 +83,103 @@ public MetricsLoggerClient(
 
   /** Log click */
-  public void logMessageClick(InAppMessage message, Action action) {
+  void logMessageClick(InAppMessage message, Action action) {
     if (!isTestCampaign(message)) {
       // If message is not a test message then log analytics
-      engagementMetricsLogger.logEvent(createEventEntry(message, CLICK_EVENT_TYPE).toByteArray());
-      logEventAsync(message, AnalyticsConstants.ANALYTICS_ACTION_EVENT, true);
+      firebaseInstallations
+          .getId()
+          .addOnSuccessListener(
+              id -> {
+                engagementMetricsLogger.logEvent(
+                    createEventEntry(message, id, CLICK_EVENT_TYPE).toByteArray());
+                logEventAsync(message, AnalyticsConstants.ANALYTICS_ACTION_EVENT, true);
+              });
     }
     // No matter what, always trigger developer callbacks
     developerListenerManager.messageClicked(message, action);
   }
 
   /** Log Rendering error */
-  public void logRenderError(InAppMessage message, InAppMessagingErrorReason errorReason) {
+  void logRenderError(InAppMessage message, InAppMessagingErrorReason errorReason) {
     if (!isTestCampaign(message)) {
       // If message is not a test message then log analytics
-      engagementMetricsLogger.logEvent(
-          createRenderErrorEntry(message, errorTransform.get(errorReason)).toByteArray());
+      firebaseInstallations
+          .getId()
+          .addOnSuccessListener(
+              id ->
+                  engagementMetricsLogger.logEvent(
+                      createRenderErrorEntry(message, id, errorTransform.get(errorReason))
+                          .toByteArray()));
     }
     // No matter what, always trigger developer callbacks
     developerListenerManager.displayErrorEncountered(message, errorReason);
   }
 
   /** Log dismiss */
-  public void logDismiss(InAppMessage message, InAppMessagingDismissType dismissType) {
+  void logDismiss(InAppMessage message, InAppMessagingDismissType dismissType) {
     if (!isTestCampaign(message)) {
       // If message is not a test message then log analytics
-      engagementMetricsLogger.logEvent(
-          createDismissEntry(message, dismissTransform.get(dismissType)).toByteArray());
-      logEventAsync(message, AnalyticsConstants.ANALYTICS_DISMISS_EVENT, false);
+      firebaseInstallations
+          .getId()
+          .addOnSuccessListener(
+              id -> {
+                engagementMetricsLogger.logEvent(
+                    createDismissEntry(message, id, dismissTransform.get(dismissType))
+                        .toByteArray());
+                logEventAsync(message, AnalyticsConstants.ANALYTICS_DISMISS_EVENT, false);
+              });
     }
   }
 
-  private CampaignAnalytics createEventEntry(InAppMessage message, EventType eventType) {
-    return createCampaignAnalyticsBuilder(message).setEventType(eventType).build();
+  private CampaignAnalytics createEventEntry(InAppMessage message, String id, EventType eventType) {
```

Comment:
nit: call it fid or instanceId to differentiate from any other type of id used in FIAM. Idk if there is one, just think it would avoid any confusion.